server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java

```diff
@@ -67,14 +68,31 @@
             CONDITIONS, ObjectParser.ValueType.OBJECT);
         PARSER.declareField((parser, request, context) -> request.createIndexRequest.settings(parser.map()),
             CreateIndexRequest.SETTINGS, ObjectParser.ValueType.OBJECT);
+        PARSER.declareField((parser, request, includeTypeName) -> {
+            if (includeTypeName) {
+                //expecting one type only
+                for (Map.Entry<String, Object> mappingsEntry : parser.map().entrySet()) {
+                    request.createIndexRequest.mapping((Map<String, Object>) mappingsEntry.getValue());
+                }
+            } else {
+                // a type is not included, add a dummy _doc type
+                Map<String, Object> mappings = parser.map();
+                if (MapperService.isMappingSourceTyped(mappings)) {
+                    throw new IllegalArgumentException("The mapping definition cannot be nested under a type " +
+                        "[" + MapperService.SINGLE_MAPPING_NAME + "] unless include_type_name is set to true.");
+                }
+                request.createIndexRequest.mapping(mappings);
+            }
+        }, CreateIndexRequest.MAPPINGS.forRestApiVersion(RestApiVersion.equalTo(RestApiVersion.V_7)), ObjectParser.ValueType.OBJECT);
         PARSER.declareField((parser, request, context) -> {
             // a type is not included, add a dummy _doc type
             Map<String, Object> mappings = parser.map();
-            if (MapperService.isMappingSourceTyped(MapperService.SINGLE_MAPPING_NAME, mappings)) {
```

Comment:
And I think we should remove this check as well - really to check properly we need to parse the mappings, and if they've included a type this will get caught further along.

Reply:
Building on what I said in a previous comment -- these were added in #38270 to catch an important case that mapping parsing wasn't able to detect. But there have been substantial improvements/ refactors to mapping parsing since then, so we may be able to remove these checks and have the tests still pass. To keep this PR a straightforward 'restore' of old logic, maybe we could look into removing these checks in a follow-up.

Reply:
Assuming I am reading the code correctly, this check is only applied when REST API compatibility for v7 is applied (i.e. CreateIndexRequest.MAPPINGS.forRestApiVersion(RestApiVersion.equalTo(RestApiVersion.V_7)), but the check is now no longer for v8+. 

I have a minor preference to keep the check when compatibility is in use, else remove the check in a follow up PR if it also requires changes to the test.

Reply:
I see, in that case I will keep the check for both 7.x and 8.x fields.
@romseygeek any views on this?

Reply:
++ let's keep it for now and I can look at removing it in a followup